sound/soc/sof/pcm.c

```diff
@@ -383,6 +383,12 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
 			 * and mark the flag to ignore the upcoming DAPM
 			 * PM events.
 			 */
+			ret = snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+			if (ret < 0) {
+				dev_warn(sdev->dev,
+					 "error %d in %s: failed to trigger KWD stream\n",
+					 ret, __func__);
+			}
```

Comment:
Sorry I don't by the argument that userspace resumes the stream, or you are relying on an application that traps the system suspend and explicitly suspends it. 

Please show a sequence explaining your findings.

Reply:
We can do this more elegantly by using a new "ignore_trigger_suspend" tuple attribute for the always running KDW FW components.  This attribute would be false for the KWD host PCM and this would be suspended as normal PCM flow, but true for the other KDW processing components. We would still trap the TRIGGER_STOP here but would continue with the TRIGGER_SUSPEND.

Reply:
@plbossart, let's start with fundamentals, if we agree on the fact that we go with `snd_sof_pcm_platform_trigger()` on suspend/resume in order to power down/up DMA then synchronization between FW-HOST is needed. Now if we do this which I believe we should we have to *care* about the resume time as this affects draining speeds. Do we agree on this?

>  Sorry I don't by the argument that userspace resumes the stream, or you are relying on an application that traps the system suspend and explicitly suspends it.

Previously we have been calling `snd_sof_pcm_platform_trigger()` in `sof_pcm_trigger` and this relies on application. In the #2285 I see you do the trigger in `hda_dsp_set_D0_state()` which is indeed independent of any user space actions but I think still slower than doing it in IPC thread. I will do tests to prove it.

Reply:
@mrajwa you can use the PR #2285 and add traces to see when the DMAs would be resumed. You just need to add traces in the IPC handler and before enabling the DMA.

But on my side I insist that your solution is currently racy and error-prone. the suspend and interrupt code are parallel threads and you will need to spend a fair amount of time working on the design to harden it with mutual exclusions, not to mention making the behavior optional and backwards compatible.

Reply:
Yes the race condition is fair point but the solution seems to be clear - power down links in `sof_pcm_trigger ` `SNDRV_PCM_TRIGGER_SUSPEND` (this is what my solution already does) and power them back in the KWD notification IPC handler. Wouldn't that fix the race condition concern? 
I have no objections to your solution, its fine for me too. Only one problem I see is the resume latency. I've just tried it locally on my volteer and the solution presented in #2285 is ~300 [ms] slower. That wouldn't be a problem for normal streams but for KWD, where you want to drain as fast as possible it is problematic.

Reply:
@mrajwa how are you measuring the ~300ms latency? Is this the time it takes to finish draining? Can you please share your BKM for measuring this?

Reply:
@ranj063, very simply - put one trace logs in `kwd_notification_handler()` and the second in `hda_dsp_set_D0_state()` (where Pierre`s solution is triggering the stream) then perform the test, and finally read dmesg time stamps and see the delta between these two.

Reply:
@mrajwa there's something very wrong here. the metric is that userspace needs to play a tone or play with a vibra in 350ms. If we already take 300ms to resume at the kernel level, well there's no need to worry really since the use case is broken. It's at least an order of magnitude larger than that's acceptable for S0ix usages.

Reply:
> Yes the race condition is fair point but the solution seems to be clear - power down links in `sof_pcm_trigger ` `SNDRV_PCM_TRIGGER_SUSPEND` (this is what my solution already does) and power them back in the KWD notification IPC handler. Wouldn't that fix the race condition concern?

As I tried to explain, you have race conditions on suspend as well. You need a handshake to make sure the DSP does not start a transfer before the DMAs are stopped.

Reply:
@mrajwa can you help provide the latency from FW keyword being detected to the draining start/done also, data like this will help comparing solutions:

Solution                                                                                       keyword detected            draining start            draining done
solution 1: without any new sync mechanism                                          t1_0                              t1_1                           t1_2
solution 2: with sync mechanism via IPC (PR#2279)                               t2_0                              t2_1                           t2_2
solution 1: with sync mechanism in suspend/resume rotine (PR#2285)  t3_0                              t3_1                           t3_2

Reply:
@plbossart, as for the race condition on suspend - if the detection happens just before the s0ix entry (before the suspend) then both solutions (in current form) have the same problem namely, the following suspend may stop the DMA/stream during draining because FW starts draining right after the detection. So what I would add to this solution would be an IPC ack from host to DSP but also we need to keep record of KWD state on host side.   States like IDLE/HOST COPY are a bare minimum. Thanks to this host will never interrupt draining and will skip suspend actions if draining is ongoing. Please note, these additional ack messages are expensive in terms of time, and it will add extra latency to the draining time. 
As I said it seems to me that your solution has the same problem, if the detection will happen right before the suspend then suspend PG action will stop the DMA anyway right?

@keyonjie, yes here are the results (time is measured from the start of draining task to its end. NOTE! Neither solution used ack from host to DSP which means draining started on its own, not waiting for any permition from host)
Solution 1: 978 ms
Solution 2: 630 ms
Solution 3: 941 ms

Also, according to dmesg the application will restart the stream after ~23 ms after we were in `hda_dsp_set_D0_state()` (this is the place where Pierres solution triggers the DMA). It means DSP will have less then 23 ms to drain ~3 seconds history buffer. We need ~10 ms to drain 2 seconds buffer so at first it doesn't look bad. However if we add ack messages then we may hit the situation when application is waiting for the data.

Reply:
> @plbossart, as for the race condition on suspend - if the detection happens just before the s0ix entry (before the suspend) then both solutions (in current form) have the same problem namely, the following suspend may stop the DMA/stream during draining because FW starts draining right after the detection. So what I would add to this solution would be an IPC ack from host to DSP but also we need to keep record of KWD state on host side. States like IDLE/HOST COPY are a bare minimum. Thanks to this host will never interrupt draining and will skip suspend actions if draining is ongoing. Please note, these additional ack messages are expensive in terms of time, and it will add extra latency to the draining time.
> As I said it seems to me that your solution has the same problem, if the detection will happen right before the suspend then suspend PG action will stop the DMA anyway right?
> 
> @keyonjie, yes here are the results (time is measured from the start of draining task to its end. NOTE! Neither solution used ack from host to DSP which means draining started on its own, not waiting for any permition from host)
> Solution 1: -
> Solution 2: 630 ms
> Solution 3: 941 ms

Thanks for the data, we can discuss based on them. How about the Solution 1 here? Did you get chance to capture data for it also.

> 
> Also, according to dmesg the application will restart the stream after ~23 ms after we were in `hda_dsp_set_D0_state()` (this is the place where Pierres solution triggers the DMA). It means DSP will have less then 23 ms to drain ~3 seconds history buffer. We need ~10 ms to drain 2 seconds buffer so at first it doesn't look bad. However if we add ack messages then we may hit the situation when application is waiting for the data.

Reply:
> @plbossart, as for the race condition on suspend - if the detection happens just before the s0ix entry (before the suspend) then both solutions (in current form) have the same problem namely, the following suspend may stop the DMA/stream during draining because FW starts draining right after the detection. So what I would add to this solution would be an IPC ack from host to DSP but also we need to keep record of KWD state on host side. States like IDLE/HOST COPY are a bare minimum. Thanks to this host will never interrupt draining and will skip suspend actions if draining is ongoing. Please note, these additional ack messages are expensive in terms of time, and it will add extra latency to the draining time.
> As I said it seems to me that your solution has the same problem, if the detection will happen right before the suspend then suspend PG action will stop the DMA anyway right?

Yes, all solutions expose this potential race on suspend. 

We need a first notification to the firmware that the DMA is going to be stopped. and we need a yes/no answer from firmware. 
If the firmware accepts the notification, then it shall not start the DMA before getting the notification that the DMA was started (be it in the IPC handler or the resume handler). 
Conversely, if the firmware rejects the notification, then the driver shall not disable the DMA.

It's really far from simple...

Reply:
I will try to come up with a better solution in following days...